ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import com.codahale.metrics.MetricRegistry;
+import com.github.ambry.config.ClusterMapConfig;
+import com.google.common.collect.Sets;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixManager;
+import org.apache.helix.HelixManagerFactory;
+import org.apache.helix.InstanceType;
+import org.apache.helix.LiveInstanceChangeListener;
+import org.apache.helix.NotificationContext;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.LiveInstance;
+import org.apache.helix.spectator.RoutingTableProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.ClusterMapUtils.*;
+
+
+/**
+ * An implementation of {@link ClusterMap} that makes use of Helix to dynamically manage the cluster information.
+ *
+ * @see <a href="http://helix.apache.org">http://helix.apache.org</a>
+ */
+class HelixClusterManager implements ClusterMap {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final String clusterName;
+  private final MetricRegistry metricRegistry;
+  private final ClusterMapConfig clusterMapConfig;
+  private final Map<String, DcZkInfo> dcToDcZkInfo = new HashMap<>();
+  private final Map<String, AmbryPartition> partitionNameToPartitionIdMap = new ConcurrentHashMap<>();
+  private final Map<String, AmbryDataNode> instanceNameToDataNodeIdMap = new ConcurrentHashMap<>();
+  private final Map<AmbryPartition, Set<AmbryReplica>> partitionIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryReplica>> dataNodeIdToReplicaIds = new ConcurrentHashMap<>();
+  private final Map<AmbryDataNode, Set<AmbryDisk>> dataNodeIdToDiskIds = new ConcurrentHashMap<>();
+  private final Map<ByteBuffer, AmbryPartition> partitionMap = new ConcurrentHashMap<>();
```

Comment:
For the running process? I don't think it would be prohibitively high. These mappings would one way or the other already exist today in the static cluster manager.